runtime/lua/vim/diagnostic.lua

```diff
@@ -813,11 +813,40 @@ local function next_diagnostic(position, search_forward, bufnr, opts, namespace)
   position[1] = position[1] - 1
   bufnr = get_bufnr(bufnr)
   local wrap = if_nil(opts.wrap, true)
-  local line_count = api.nvim_buf_line_count(bufnr)
+
+  local auto_severity = opts.severity == 'auto'
+  if auto_severity then
+    -- Clear the severity filter so that we get all diagnostics
+    opts.severity = nil
+  end
+
   local diagnostics =
     get_diagnostics(bufnr, vim.tbl_extend('keep', opts, { namespace = namespace }), true)
+
+  -- When 'auto' severity is set, sort the diagnostics by severity. The first diagnostic then
+  -- contains the highest severity, and we can discard all diagnostics with a lower severity.
+  if auto_severity then
+    table.sort(diagnostics, function(a, b)
+      return a.severity < b.severity
+    end)
+
+    -- Find the first diagnostic where the severity does not match the highest severity, and remove
+    -- that element and all subsequent elements from the array
+    local worst = diagnostics[1].severity
```

Comment:
```suggestion
    local worst = (diagnostics[1] or {}).severity
```

I think otherwise this fails with no diagnostics

Reply:
81fa7f649d80c3bb8e601176cb1dbf3e2771fa85